packages/frontend/src/providers/SwapHistoryProvider.tsx

```diff
@@ -0,0 +1,84 @@
+import React, { createContext, useReducer, useContext, useEffect, useState } from 'react';
+import { Quote } from '@sifi/sdk';
+import { SwapHistoryModal } from 'src/modals/SwapHistoryModal/SwapHistoryModal';
+
+type SwapEvent = {
+  quote: Quote;
+  createdAt: Date;
+  hash: `0x${string}`;
+};
+
+type SwapHistoryState = {
+  swapHistory: SwapEvent[];
+};
+
+type Action = { type: 'ADD_SWAP_EVENT'; payload: SwapEvent };
+
+const initialState: SwapHistoryState = {
+  swapHistory: (JSON.parse(localStorage.getItem('swapHistory') || '[]') as SwapEvent[]).map(
+    event => ({
+      ...event,
+      createdAt: new Date(event.createdAt),
+    })
+  ),
+};
+
+const swapHistoryReducer = (state: SwapHistoryState, action: Action): SwapHistoryState => {
+  switch (action.type) {
+    case 'ADD_SWAP_EVENT':
+      return { ...state, swapHistory: [action.payload, ...state.swapHistory] };
+    default:
+      return state;
+  }
+};
+
+const SwapHistoryContext = createContext<{
+  state: SwapHistoryState;
+  dispatch: React.Dispatch<Action>;
+  toggleHistoryModal: () => void;
+}>({
+  state: initialState,
+  dispatch: () => undefined,
+  toggleHistoryModal: () => undefined,
+});
+
+const SwapHistoryProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
+  const [state, dispatch] = useReducer(swapHistoryReducer, undefined, () => {
+    const storedState = localStorage.getItem('swapHistory');
```

Comment:
Minor comment, maybe we store this key as a variable.

Looking good overall though.